src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilter.java

```diff
@@ -0,0 +1,184 @@
+import com.puppycrawl.tools.checkstyle.xpath.AbstractNode;
+import com.puppycrawl.tools.checkstyle.xpath.RootNode;
+import net.sf.saxon.om.Item;
+import net.sf.saxon.sxpath.XPathDynamicContext;
+import net.sf.saxon.sxpath.XPathEvaluator;
+import net.sf.saxon.sxpath.XPathExpression;
+import net.sf.saxon.trans.XPathException;
+
+/**
+ * This filter processes {@link TreeWalkerAuditEvent}
+ * objects based on the criteria of file, check, module id, xpathQuery.
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathFilter implements TreeWalkerFilter {
+    /** The regexp to match file names against. */
+    private final Pattern fileRegexp;
+
+    /** The pattern for file names. */
+    private final String filePattern;
+
+    /** The regexp to match check names against. */
+    private final Pattern checkRegexp;
+
+    /** The pattern for check class names. */
+    private final String checkPattern;
+
+    /** Module id filter. */
+    private final String moduleId;
+
+    /** Xpath expression. */
+    private final XPathExpression xpathExpression;
+
+    /** Xpath query. */
+    private final String xpathQuery;
+
+    /**
+     * Creates a {@code XpathElement} instance.
+     * @param files regular expression for names of filtered files
+     * @param checks regular expression for filtered check classes
+     * @param moduleId the module id
+     * @param query the xpath query
+     */
+    public XpathFilter(String files, String checks,
+                       String moduleId, String query) {
+        filePattern = files;
+        fileRegexp = Pattern.compile(files);
+        checkPattern = checks;
+        if (checks == null) {
+            checkRegexp = null;
+        }
+        else {
+            checkRegexp = CommonUtils.createPattern(checks);
+        }
+        this.moduleId = moduleId;
+        xpathQuery = query;
+        if (xpathQuery == null) {
+            xpathExpression = null;
+        }
+        else {
+            final XPathEvaluator xpathEvaluator = new XPathEvaluator();
+            try {
+                xpathExpression = xpathEvaluator.createExpression(xpathQuery);
+            }
+            catch (XPathException ex) {
+                throw new IllegalStateException("Unexpected xpath query", ex);
+            }
+        }
+    }
+
+    @Override
+    public boolean accept(TreeWalkerAuditEvent event) {
+        return !isFileNameAndModuleAndCheckNameMatching(event)
+                || !isXpathQueryMatching(event);
+    }
+
+    /**
+     * Is matching by file name, moduleId and Check name.
+     * @param event event
+     * @return true if it is matching
+     */
+    private boolean isFileNameAndModuleAndCheckNameMatching(TreeWalkerAuditEvent event) {
+        return event.getFileName() != null
+                && fileRegexp.matcher(event.getFileName()).find()
+                && event.getLocalizedMessage() != null
+                && (moduleId == null || moduleId.equals(event.getModuleId()))
```

Comment:
In defense of the current implementation

What if `id` attribute of the `suppress` tag would be responsible for environment. Suppose we have two environments: **test** and **prod**. And we want to specify different suppressions for source and test files.

The following example illustrates how different suppressions can coexist inside the same file 

```

<suppress 
checks="(RightCurly|LeftCurly)" id="test" files="Test.java" lines="1-100" columns="4"/>

<suppress 
checks="RightCurly" id="prod" files="MainController.java" lines="4" columns="4"/>
```